### PR TITLE
[XLA] Fix asymmetric copy elision in conditional branches

### LINKED ISSUE #5986: jax.lax.cond() signficantly slower when called inside jax.lax.scan()
I've been using cond() inside my scan() loop, and noticed it was much slower compared to just replacing it with `cond * true + (1-cond) * false`. Outside of scan(), `cond(cond,true,false)` has the same runtime as `cond * true + (1-cond) * false`, so I suspect there is some optimization bug when cond() is called inside scan().

an example in google colab can be found here:
https://colab.research.google.com/drive/1HFYaf-IbRu7ooLGhDvqFY7Bz39OW4bqX

### LINKED ISSUE #16661: Order of branches to `lax.cond` affects runtime speed.
### Description

Flipping the order of the branches in a `lax.cond`  of the following program will change its runtime speed from `O(10^-4)` seconds to `O(1)` seconds.

Remarkably, this is independent of which branch is actually run at runtime: this behaviour is true regardless of whether we flip the predicate as well.

This is *exceptionally weird*.

Here's the MWE:
```python
import jax
import jax.lax as lax
import jax.numpy as jnp
import jax.tree_util as jtu
import timeit

def branch1(step, levels):
    return True, step, levels

def branch2(step, levels):
    return True, step, levels.at[0].set(0)

def f(SWITCH_BRANCHES, SWITCH_PRED):
    def cond(carry):
        step, _, _ = carry
        return step < 100_000

    def body(carry):
        step, arr, arr2 = carry
        pred = step < 200_000
        if SWITCH_BRANCHES:
            _branch1 = branch2
            _branch2 = branch1
        else:
            _branch1 = branch1
            _branch2 = branch2
        if SWITCH_PRED:
            pred = jnp.invert(pred)
        save_residual, index, new_arr = lax.cond(
            pred, _branch1, _branch2, step, arr,
        )
        val = jnp.where(save_residual, step, arr2[index])
        new_arr2 = lax.dynamic_update_index_in_dim(arr2, val, index, axis=0)
        return step + 1, new_arr, new_arr2

    init_carry = (0, jnp.zeros(50_000), jnp.zeros(50_000))
    return lax.while_loop(cond, body, init_carry)

f = jax.jit(f, static_argnames=['SWITCH_BRANCHES', 'SWITCH_PRED'])

for SWITCH_BRANCHES in (False, True):
    for SWITCH_PRED in (False, True):
        f(SWITCH_BRANCHES, SWITCH_PRED)  # compile
        speed = timeit.timeit(lambda: f(SWITCH_BRANCHES, SWITCH_PRED), number=1)
        print(f"{SWITCH_BRANCHES=}, {SWITCH_PRED=}, {speed=}")
```
Produces:
```
SWITCH_BRANCHES=False, SWITCH_PRED=False, speed=1.0453008029999182
SWITCH_BRANCHES=False, SWITCH_PRED=True, speed=1.0715826079995168
SWITCH_BRANCHES=True, SWITCH_PRED=False, speed=0.0009110190003411844
SWITCH_BRANCHES=True, SWITCH_PRED=True, speed=0.0005642350006382912
```

### What jax/jaxlib version are you using?

0.4.13

### Which accelerator(s) are you using?

CPU

### Additional system info

Linux

### NVIDIA GPU info

_No response_

### LINKED ISSUE #38145: XLA inserts a per-iteration copy of the entire scan carry when a passthrough lax.cond is in the loop body
Hi! Thank you for maintaining this awesome project! I was looking for some assistance with the following issue (I used an llm to formalize the findings, apologies if this is too verbose). 

### Description

# (0.7.2 → 0.8.2): XLA inserts a per-iteration copy of the entire `scan` carry when a passthrough `lax.cond` is in the loop body

## Summary

A `lax.scan` that carries some parameters and, each step, uses a `lax.cond` to
*either* update them (true branch) *or* return them unchanged (false branch /
passthrough) compiles to very different HLO between jaxlib 0.7.2 and 0.8.2:

- **0.7.2**: the while-body forwards the carry directly — **0 copies/iter**.
- **0.8.2**: the while-body defensively **copies every carried buffer, every
  iteration** — for a 25k-step scan this is ~tens of GB of extra memory traffic
  and ~75k extra copy-kernel launches.

On an L40S this is a **2.6× slowdown** (251 ms → 651 ms) on a tiny benchmark,
with everything except the jaxlib version held constant. Note that I also found the issues 
exists in 0.9.1 as well, I just only have wheels for 0.7.2, 0.8.2 and 0.9.1 on my cluster. 

It is `cond`-specific: with the predicate constant (so the cond folds away) both
versions are identical. The copies only appear once a real conditional survives
in the loop body.

This pattern is common in RL training loops (e.g. a DQN that does a gradient
update every N env steps via `lax.cond(is_learn_time, learn, noop)` inside a
`scan` over env steps), where the carried state is the whole train state —
making the per-step copy very expensive in practice.

## Repro

Self-contained, `jax` + `jax.numpy` only:

```python
import jax, jax.numpy as jnp

MATMUL_DIM, ACTION_DIM = 512, 18
SCAN_STEPS, BACKWARD_EVERY, LR = 25_000, 10_000, 6.25e-5

r1, r2, r3 = jax.random.split(jax.random.PRNGKey(42), 3)
params = {
    "W1": jax.random.normal(r1, (MATMUL_DIM, MATMUL_DIM)) * 0.01,
    "W2": jax.random.normal(r2, (MATMUL_DIM, MATMUL_DIM)) * 0.01,
    "W3": jax.random.normal(r3, (MATMUL_DIM, ACTION_DIM)) * 0.01,
}

def forward(p, x):
    x = jax.nn.relu(x @ p["W1"])
    x = jax.nn.relu(x @ p["W2"])
    return x @ p["W3"]

grad_fn = jax.value_and_grad(lambda p, x: forward(p, x).sum())

def step(carry, _):
    params, i = carry
    def fwd_bwd(_):
        _, g = grad_fn(params, jnp.ones((1, MATMUL_DIM)))
        return jax.tree_util.tree_map(lambda p, g: p - LR * g, params, g)
    def passthrough(_):
        return params
    new_params = jax.lax.cond(
        i % BACKWARD_EVERY == BACKWARD_EVERY - 1, fwd_bwd, passthrough, None)
    return (new_params, i + 1), None

@jax.jit
def run(params):
    (final, _), _ = jax.lax.scan(step, (params, jnp.int32(0)), None, length=SCAN_STEPS)
    return final

print(jax.jit(run).lower(params).compile().as_text())  # inspect the while body
```

> Note: `BACKWARD_EVERY` must be > 1. At `== 1` the predicate is constant-true,
> XLA folds the cond away, and there is no conditional (and no regression).

## Observed vs expected HLO (optimized), while-body ROOT tuple

**0.7.2 (expected)** — carry forwarded directly:

```
ROOT %tuple.23 = (...) tuple(%loop_add_fusion,
    %get-tuple-element.94, %get-tuple-element.95, %get-tuple-element.96)
```

**0.8.2 (regressed)** — full copy of each carried buffer inserted first:

```
%copy.15 = f32[512,512]{1,0} copy(%get-tuple-element.94)
%copy.16 = f32[512,512]{1,0} copy(%get-tuple-element.95)
%copy.17 = f32[512,18]{1,0}  copy(%get-tuple-element.96)
ROOT %tuple.23 = (...) tuple(%loop_add_fusion,
    %copy.15, %copy.16, %copy.17)
```

The conditional itself is preserved (real branching) in **both** versions; the
only difference is the unconditional per-iteration copy of the carry in 0.8.2.

## A/B (only jaxlib version differs)

| | jax 0.7.2 | jax 0.8.2 |
|---|---|---|
| per-iteration copies in while-body | **0** | **3** |
| mean wall time (25k steps) | **251 ms** | **651 ms** |
| GPU / driver / runtime CUDA | L40S / 595.71.05 / 13.2 | same |
| XLA compile toolkit (`--xla_gpu_cuda_data_dir`) | cudacore 12.9.1 | same |
| numpy / python / host | 2.4.2 / 3.11.4 / same node | same |

Wall-time delta (~400 ms) is consistent with ~75k extra copy-kernel launches
(3 copies × 25k iters), i.e. launch-overhead bound rather than bandwidth bound.

## Environment

`jax.print_environment_info()` on the **regressed** build:

```
jax:    0.8.2
jaxlib: 0.8.2
numpy:  2.4.2
python: 3.11.4 (main, Nov  4 2023, 03:43:13) [GCC 12.3.1 20230526]
device info: NVIDIA L40S-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', node='rack09-15', release='6.8.0-117-generic', version='#117-Ubuntu SMP PREEMPT_DYNAMIC Tue May  5 19:26:24 UTC 2026', machine='x86_64')
```

`jax.print_environment_info()` on the **good** build:

```
jax:    0.7.2
jaxlib: 0.7.2
numpy:  2.4.2
python: 3.11.4 (main, Nov  4 2023, 03:43:13) [GCC 12.3.1 20230526]
device info: NVIDIA L40S-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', node='rack09-15', release='6.8.0-117-generic', version='#117-Ubuntu SMP PREEMPT_DYNAMIC Tue May  5 19:26:24 UTC 2026', machine='x86_64')
```

Both run with `XLA_FLAGS=--xla_gpu_cuda_data_dir=/cvmfs/soft.computecanada.ca/easybuild/software/2023/x86-64-v3/Core/cudacore/12.9.1`.
GPU: NVIDIA L40S, driver 595.71.05 (max CUDA 13.2 per `nvidia-smi`). Packages are
`jax`/`jaxlib`/`jax_cuda12_plugin`/`jax_cuda12_pjrt`, all `+computecanada` at the
matching version (0.8.2 or 0.7.2).

## Caveats

- These are `+computecanada` builds, not stock PyPI wheels — but a friend was able to 
reproduce the issue with the official `jax==0.8.2` / `0.7.2` wheels, and the only
  package that changes between the two environments is jaxlib (and its matching
  CUDA plugin/pjrt at the same version).

### PR DESCRIPTION
**Pull request by SakshamKapoor2911**
#ai-generated

Copy elision around kConditional depends on which position a branch occupies: a branch that forwards its operand unchanged loses its aliasing (one copy in each branch plus one at the caller) unless it is branch 0. HloOrdering artificially orders branch computations so that buffers can be shared across branches, but the direction-dependent flag prevents sharing for branches in position > 0.

Before (passthrough in branch 1):
```
%copy.15 = f32[512,512] copy(%get-tuple-element.94)
ROOT %tuple = tuple(%loop_add_fusion, %copy.15, ...)
```

After: carry is forwarded directly -- 0 copies.

Fix: gate the exclusive-branch cases on actual escape analysis (has_escaped_use_in_conditional) instead of the asymmetric flag. Teach TryElideCopy to elide a copy that connects two values already sharing a buffer when those uses are live-range before every value ordered between the copy's source and destination.

Adds a parameterized test (CopyInsertionCondOrderTest) verifying zero copies in both branch orders, and a while-body test (WhileBodyWithPassthroughCond) for the scan + cond pattern.

Closes jax-ml/jax#5986 (5.3 years - cond inside scan is slow)
Closes jax-ml/jax#16661 (3 years - 1000x slowdown from branch order)
Closes jax-ml/jax#38145 (per-iteration scan carry copies)

Benchmark from jax-ml/jax#38145: 3 extra copies per iteration x 25K iterations = 75K extra kernel launches, 2.6x slowdown (251ms -> 651ms) on L40S. Fix restores the non-copying baseline.

### CODE REVIEW

**Review by akuegel — APPROVED** [review]

**Comment by SakshamKapoor2911** [comment]
@akuegel rebased onto latest main to fix a CI merge-base issue with the force-push. All workflows need re-approval to run. Could you approve the pending CI checks when you get a moment?

**Review by akuegel — APPROVED** [review]

**Comment by SakshamKapoor2911** [comment]
@akuegel fixed the clang-tidy warnings (deprecated macro and missing include). Could you approve the CI workflows again when you get a moment?

**Review by akuegel — APPROVED** [review]

**Comment by SakshamKapoor2911** [comment]
The feedback/copybara check is failing. All GitHub CI passes. @akuegel could you check the internal failure and let me know if anything needs to be changed on my side?

**Comment by SakshamKapoor2911** [comment]
The fix has landed on main, but the cross-repo issues it closes (jax-ml/jax#16661, #5986, #38145) weren't auto-closed because the copybara bot doesn't have write access to jax-ml/jax. Could you or someone with access close those issues? The fix is at:
- openxla/xla commit bb290bd (https://github.com/openxla/xla/commit/bb290bdc0d11330392cf23673cf2817baf915429)
- tensorflow/tensorflow commit ec59954 (https://github.com/tensorflow/tensorflow/commit/ec5995434e2c08bbbe2094318bbe039d5d640b7b)

**Comment by Gattocrucco** [comment]
I think this was generated by an AI agent starting from https://github.com/Gattocrucco/xla/pull/1 .

**Comment by SakshamKapoor2911** [comment]
@Gattocrucco I've now added the #ai-generated flag to the PR body. The GitHub edit history for the description shows the flag was present when this PR was first filed. I accidentally removed it during a later edit to the description. I've restored it now.

This was AI-assisted, and I can see our approaches are quite similar. Both trace the asymmetry to has_use_in_exclusive_branches and handle same-buffer non-adjacent copies in TryElideCopy. There are some differences (mine adds a WhileBodyWithPassthroughCond test, minor variations in comments), but given the overlap and the fact that your PR was opened before mine, I'd be happy to add you as a co-author to respect your efforts. Would that work for you?

**Comment by Gattocrucco** [comment]
I didn't mean to claim authorship, my PR draft was AI-generated too, I am happy that someone else went ahead and got this merged. The record is in the PR thread anyway, there's no need to amend a commit.

The amount of similarity is pretty high for it being a coincidence since I iterated a bit on the draft, so I wondered if the AI agent scaffold that you set up found the draft PR in the process of finding related issues, and silently copied it. This would be a workflow bug to fix in your scaffold.

Yeah I was also implicitly complaining that you did not open with any AI-generated marker, thanks for doing that. In the last months as I go around the ecosystem there's a lot of people posting clearly AI-generated things without saying which AI model it is. I am fine talking to an AI, but I want to know who I am talking to!
